src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -529,6 +568,97 @@ private CachingAmazonDynamoDbStreams(AmazonDynamoDBStreams amazonDynamoDbStreams
+     * Gets the {@code DescribeStreamResult} from the DescribeStream API.
+     * @param describeStreamRequest Describe stream request.
+     * @return Stream details for the given request with all the shards currently available.
+     */
+    protected DescribeStreamResult loadStreamDescriptionForAllShards(DescribeStreamRequest describeStreamRequest)
+        throws AmazonDynamoDBException {
+
+        List<Shard> allShards = new ArrayList<>();
+        String streamArn = describeStreamRequest.getStreamArn();
+        String lastShardId = describeStreamRequest.getExclusiveStartShardId();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("cache miss in describe stream cache for key: {}", streamArn);
+        }
+        // TODO metrics: count cache miss
+
+        DescribeStreamResult result = null;
+        DescribeStreamResult previousResult;
+        do {
+            describeStreamRequest = new DescribeStreamRequest().withStreamArn(streamArn)
+                .withExclusiveStartShardId(lastShardId);
+            previousResult = result;
+            result = super.describeStream(describeStreamRequest);
+
+            if (result != null && result.getStreamDescription().getShards() != null) {
+                List<Shard> shards = result.getStreamDescription().getShards();
+                allShards.addAll(shards);
+                lastShardId = result.getStreamDescription().getLastEvaluatedShardId();
+            } else {
+                // since the request was null, update lastShardId to null since we can't get the next request
+                // from a null response
+                lastShardId = null;
+            }
+        } while (lastShardId != null);
+
+        if (result != null) {
+            StreamDescription streamDescription = result.getStreamDescription();
+            streamDescription.setShards(allShards);
+            result.setStreamDescription(streamDescription);
+        } else if (previousResult != null) {
+            StreamDescription streamDescription = previousResult.getStreamDescription();
+            streamDescription.setShards(allShards);
+            result = previousResult;
+            result.setStreamDescription(streamDescription);
+        } else {
+            result = new DescribeStreamResult().withStreamDescription(new StreamDescription().withStreamArn(streamArn)
+                .withShards(allShards));
+        }
+
+        return result;
+    }
+
+    /**
+     * Gets {@code DescribeStreamResult} for request from cache or describeStream API.
+     *
+     * @param describeStreamRequest Describe stream request.
+     * @return {@code DescribeStreamResult}.
+     */
+    @Override
+    public DescribeStreamResult describeStream(DescribeStreamRequest describeStreamRequest) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("describeStream request={}", describeStreamRequest);
+        }
+        DescribeStreamResult result = lookupDescribeStreamResult(describeStreamRequest);
```

Comment:
What is the value in having separate `describeStream` and `lookupDescribeStreamResult` methods?

Reply:
simplified in beed68d0c791549292ac4b01bf75b5b860af4ad6